tidy3d/plugins/smatrix/component_modelers/terminal.py

```diff
@@ -0,0 +1,425 @@
+"""Tool for generating an S matrix automatically from a Tidy3d simulation and lumped port definitions."""
+from __future__ import annotations
+
+from typing import Tuple, Dict
+
+import pydantic.v1 as pd
+import numpy as np
+import xarray as xr
+
+from ....constants import C_0, fp_eps
+from ....components.simulation import Simulation
+from ....components.geometry.utils_2d import snap_coordinate_to_grid
+from ....components.monitor import FieldMonitor
+from ....components.source import GaussianPulse, UniformCurrentSource
+from ....components.data.sim_data import SimulationData
+from ....components.types import Ax
+from ....components.viz import add_ax_if_none, equal_aspect
+from ....components.base import cached_property
+from ....components.lumped_element import LumpedResistor
+from ....exceptions import ValidationError
+from ....web.api.container import BatchData
+
+from .base import AbstractComponentModeler, FWIDTH_FRAC
+from ..ports.lumped import LumpedPortDataArray, LumpedPort
+
+
+class TerminalComponentModeler(AbstractComponentModeler):
+    """Tool for modeling two-terminal multiport devices and computing port parameters
+    with lumped ports."""
+
+    ports: Tuple[LumpedPort, ...] = pd.Field(
+        (),
+        title="Lumped ports",
+        description="Collection of lumped ports associated with the network. "
+        "For each port, one simulation will be run with a lumped port source.",
+    )
+
+    @equal_aspect
+    @add_ax_if_none
+    def plot_sim(
+        self, x: float = None, y: float = None, z: float = None, ax: Ax = None, **kwargs
+    ) -> Ax:
+        """Plot a :class:`Simulation` with all sources added for each port, for troubleshooting."""
+
+        plot_sources = []
+        for port_source in self.ports:
+            source_0 = self.to_source(port=port_source, snap_center=None)
+            plot_sources.append(source_0)
+        sim_plot = self.simulation.copy(update=dict(sources=plot_sources))
+        return sim_plot.plot(x=x, y=y, z=z, ax=ax, **kwargs)
+
+    @equal_aspect
+    @add_ax_if_none
+    def plot_sim_eps(
+        self, x: float = None, y: float = None, z: float = None, ax: Ax = None, **kwargs
+    ) -> Ax:
+        """Plot permittivity of the :class:`Simulation` with all sources added for each port."""
+
+        plot_sources = []
+        for port_source in self.ports:
+            source_0 = self.to_source(port=port_source, snap_center=None)
+            plot_sources.append(source_0)
+        sim_plot = self.simulation.copy(update=dict(sources=plot_sources))
+        return sim_plot.plot_eps(x=x, y=y, z=z, ax=ax, **kwargs)
+
+    def to_source(self, port: LumpedPort, snap_center: float) -> UniformCurrentSource:
+        """Create a current source from a given lumped port."""
+        # The source is expanded slightly to make sure that, when discretized, it covers the
```

Comment:
Could you remind me where the source is slightly expanded, I don't find that in this function?

Reply:
I think that is an old comment, I don't see any place where that occurs.